4337/contracts/Safe4337Module.sol

```diff
@@ -199,62 +158,147 @@ contract Safe4337Module is IAccount, HandlerContext, CompatibilityFallbackHandle
+     * @param userOp The ERC-4337 user operation.
+     * @return operationData Encoded operation data bytes.
      * @return validAfter The timestamp the user operation is valid from.
      * @return validUntil The timestamp the user operation is valid until.
-     * @return signature The actual signature for the Safe user operation.
+     * @return signatures The Safe signatures extracted from the user operation.
      */
-    function _splitSignatureData(
-        bytes calldata signatureData
-    ) internal pure returns (uint48 validAfter, uint48 validUntil, bytes calldata signature) {
-        validAfter = uint48(bytes6(signatureData[0:6]));
-        validUntil = uint48(bytes6(signatureData[6:12]));
-        signature = signatureData[12:];
+    function _getSafeOp(
+        UserOperation calldata userOp
+    ) internal view returns (bytes memory operationData, uint48 validAfter, uint48 validUntil, bytes calldata signatures) {
+        {
+            bytes calldata sig = userOp.signature;
+            validAfter = uint48(bytes6(sig[0:6]));
+            validUntil = uint48(bytes6(sig[6:12]));
+            signatures = sig[12:];
+        }
+
+        // It is important that **all** user operation fields are represented in the `SafeOp` data somehow, to prevent
+        // user operations from being submitted that do not fully respect the user preferences. The only exception are
+        // the `signature` bytes. Note that even `initCode` needs to be represented in the operation data, otherwise
+        // it can be replaced with a more expensive initialization that would charge the user additional fees.
+        {
+            bytes32 structHash;
+            bytes32 typeHash = SAFE_OP_TYPEHASH;
+            bytes32 initCodeHash = keccak256(userOp.initCode);
+            bytes32 callDataHash = keccak256(userOp.callData);
+            bytes32 paymasterAndDataHash = keccak256(userOp.paymasterAndData);
+            address entryPoint = SUPPORTED_ENTRYPOINT;
+
+            // Use assembly to work around Solidity "stack too deep" errors.
+            // solhint-disable-next-line no-inline-assembly
+            assembly ("memory-safe") {
+                let ptr := mload(0x40)
+
+                mstore(ptr, typeHash)
+                calldatacopy(add(ptr, 32), userOp, 384)
+                mstore(add(ptr, 96), initCodeHash)
+                mstore(add(ptr, 128), callDataHash)
+                mstore(add(ptr, 320), paymasterAndDataHash)
+                mstore(add(ptr, 352), validAfter)
+                mstore(add(ptr, 384), validUntil)
+                mstore(add(ptr, 416), entryPoint)
+
+                structHash:= keccak256(ptr, 448)
+            }
+
+            operationData = abi.encodePacked(
+                bytes1(0x19),
+                bytes1(0x01),
+                domainSeparator(),
+                structHash
+            );
+        }
+    }
+
+    /*
+    function _getSafeOpLessAssembly(
```

Comment:
This is an alternative implementation that consumes ~600 - ~1000 more gas per user operation recovery. Is is the “code” version of what we are trying to accomplish with the assembly ☝️.